book-content/chapters/13-modules-scripts-declaration-files.md

```diff
@@ -0,0 +1,988 @@
+
+## Declaration Files
+
+Declaration files are files in TypeScript which have a special extension: `.d.ts`. These files are used for two main purposes in TypeScript: describing JavaScript code, and adding types to the global scope. We'll explore both below.
+
+### Declaration Files Describe JavaScript
+
+Let's say part of our codebase is written in JavaScript, and we want to keep it that way. We have a `musicPlayer.js` file that exports a `playTrack` function:
+
+```javascript
+// musicPlayer.js
+
+export const playTrack = (track) => {
+  // Complicated logic to play the track...
+  console.log(`Playing: ${track.title}`);
+};
+```
+
+If we try to import this file into a TypeScript file, we'll get an error:
+
+```ts twoslash
+// @errors: 2307
+// inside of app.ts
+
+import { playTrack } from "./musicPlayer";
+```
+
+This error occurs because TypeScript doesn't have any type information for the `musicPlayer.js` file. To fix this, we can create a declaration file with the same name as the JavaScript file, but with a `.d.ts` extension:
+
+```typescript
+// musicPlayer.d.ts
+export function playTrack(track: {
+  title: string;
+  artist: string;
+  duration: number;
+}): void;
+```
+
+It's important to notice that this file doesn't contain any implementation code. It only describes the types of the functions and variables in the JavaScript file.
+
+Now, when we import the `musicPlayer.js` file into a TypeScript file, the error will be resolved, and we can use the `playTrack` function as expected:
+
+```typescript
+// inside of app.ts
+
+import { playTrack } from "./musicPlayer";
+
+const track = {
+  title: "Otha Fish",
+  artist: "The Pharcyde",
+  duration: 322,
+};
+
+playTrack(track);
+```
+
+Types and interfaces can also be declared and exported in declaration files:
+
+```tsx
+// inside of musicPlayer.d.ts
+export interface Track {
+  title: string;
+  artist: string;
+  duration: number;
+}
+
+export function playTrack(track: Track): void;
+```
+
+Just like in `.ts` files, these can also be imported and used in other TypeScript files:
+
+```tsx
+// inside of app.ts
+
+import { Track, playTrack } from "./musicPlayer";
+```
+
+It's important to note that declaration files are not checked against the JavaScript files they describe. We can very easily make a mistake in our declaration file, such as changing `playTrack` to `playTRACK`, and TypeScript won't complain.
+
+So, describing JavaScript files by hand can be error-prone - and not usually recommended.
+
+### Declaration Files Can Add To The Global Scope
+
+Just like regular TypeScript files, declaration files can be treated as either modules or scripts based on whether or not the `export` keyword is used. In the example above, `musicPlayer.d.ts` is treated as a module because it includes the `export` keyword.
```

Comment:
I'm honestly not sure if `export` is enough for declaration files to be treated as modules. But maybe it is and the effect I'm going to show is caused by something else...

Despite `export` being present in the declaration file you can still import just *anything* available identifier from it: [TS playground](https://www.typescriptlang.org/play/?filetype=d.ts#code/C4TwDgpgBAYg9nKBeKBvAUFLUBmCBcUAzsAE4CWAdgOboC+66EAHmHKcFKJFAEICGpZGkzYARoMKUArgFsxEUvUYATCAGMANoOg5pldcHJxKXCCQAUASkIA3OORWNysthzSwEUfkSgAlCEp+WQgVeEQ6XFI4WSgAcgA6AHoqMGlgOPQkpOxcgD0AfmdXdk5UMxJvX1JA4NCAFXNOSJxo2MSUyjSMrJzcrEKgA). It doesn't feel very modele-y ;p

You can "fix" it by including `export {}` in the file: [TS playground](https://www.typescriptlang.org/play/?filetype=d.ts#code/C4TwDgpgBAYg9nKBeKBvAUFLUBmCBcUAzsAE4CWAdgOboC+66EAHmHKcFKJFAEICGpZGkzYARoMKUArgFsxEUvUYATCAGMANoOg5pldcHJxKXCCQAUASkIA3OORWMWbDmjoBuRuVmvOqWAQofiIoACUISn5ZCBV4RDpcUjhZKAByADoAeiowaWA09Cys7FKAPQB+b192fzMSYNDSSOjYgBVzTkScZNTMnMo8gqKS0qxKoA)